src/lib/log.h

```diff
@@ -0,0 +1,44 @@
+#ifndef __SWUPD_LOG__
+#define __SWUPD_LOG__
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// int to string for macros
+#define STR2(val) #val
+#define STR(val) STR2(val)
+
+#define LOG_ALWAYS 0
+#define LOG_ERROR 2
+#define LOG_WARN 4
+#define LOG_INFO 6
+#define LOG_DEBUG 8
+
+/*
+ * Set the minimum priority log to be printed by log functions.
+ */
+void set_log_level(int log_level);
+
+/*
+ * Print a message to 'out' if 'log_level' is greater or equal the log level
+ * set by set_log_level().
+ */
+void log_full(int log_level, FILE *out, const char *str, ...);
+
+/*
+ * Helpers for logging: print(), error(), warn(), info(), debug()
+ */
+#define print(_fmt, ...) log_full(LOG_ALWAYS, stdout, _fmt, ##__VA_ARGS__);
+#define error(_fmt, ...) log_full(LOG_ERROR, stderr, "Error: " _fmt, ##__VA_ARGS__);
+#define warn(_fmt, ...) log_full(LOG_WARN, stderr, "Warning: " _fmt, ##__VA_ARGS__);
+#define info(_fmt, ...) log_full(LOG_INFO, stdout, _fmt, ##__VA_ARGS__);
+#define debug(_fmt, ...) log_full(LOG_DEBUG, stdout, "DEBUG (" __FILE__ ":" STR(__LINE__) "): " _fmt, ##__VA_ARGS__);
```

Comment:
What about adding the time to the DEBUG output? couldn't that be useful?

Reply:
Nice idea. I had a different idea now about the log names. I'll rename info() to msg() and debug() to info(). and print "INFO ()...." instead of debug. And rename --debug to --verbose. What do you think?

So --quiet will show print() and error()
Standard will show print(), error(), warn(), msg()
--verbose will show print(), error(), warn() , msg(), info()

Is this better?

Reply:
well aren't debug, info, warning, error more standard?

Reply:
ok. I'll leave the way it it then.
--quiet shows print() and error()
standard shows print(), error(), warn(), info()
--debug shows print(), error(), warn(), info(), debug()